### PR TITLE
GnarlyGenotyper tech debt

### LINKED ISSUE #5988: `-ERC` marked “beta-feature” in haplotypecaller 4.1.0.0 and later
## Bug Report
`-ERC` has been marked as a “beta-feature” in the tool docs ever since gatk4.1.0.0.
### Affected tool(s) or class(es)
HaplotypeCaller
### Affected version(s)
v4.1.0.0 and later
### Description 
https://software.broadinstitute.org/gatk/documentation/tooldocs/4.1.0.0/org_broadinstitute_hellbender_tools_walkers_haplotypecaller_HaplotypeCaller.php#--emit-ref-confidence

#### Expected behavior
It should be beta for M2 and non-beta / regular for HC. Jonn mentioned that, when it was moved to a common args class it was marked as beta because for M2 it is, hence it's confusing

----

### PR DESCRIPTION
**Pull request by ldgauthier**
Addressing @droazen 's outstanding comments from #4947 

Also closes #5988

### CODE REVIEW

**Comment by droazen** [comment]
@ldgauthier Could you please rebase this branch onto the latest master, so that we can see whether tests pass?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AnnotationUtils.java [hunk 1]
```diff
@@ -50,16 +50,7 @@ public static String encodeAnyASList( final List<?> somethingList) {
...
-            return true;
-        }
-        if (annotation instanceof AS_StrandBiasTest) {
-            return true;
-        }
-        if (annotation instanceof AS_RMSMappingQuality) {
-            return true;
-        }
-        if (annotation instanceof AS_StandardAnnotation) {
+        if (annotation instanceof AlleleSpecificAnnotation) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AnnotationUtils.java]: This branch references the new `AlleleSpecificAnnotation` interface in several places, but does not seem to actually introduce it. Maybe you just forgot to `git add` it?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java [hunk 2]
```diff
@@ -273,6 +273,18 @@ private void validateAndInitializeArgs() {
...
+        if (hcArgs.floorBlocks && !emitReferenceConfidence()) {
+            throw new UserException(HaplotypeCallerArgumentCollection.OUTPUT_BLOCK_LOWER_BOUNDS + " refers to GVCF blocks," +
+                    " so reference confidence mode (" + AssemblyBasedCallerArgumentCollection.EMIT_REFERENCE_CONFIDENCE_LONG_NAME +
+                    ") must be specified.");
+        }
+
+        if (hcArgs.floorBlocks && !emitReferenceConfidence()) {
+            throw new UserException(HaplotypeCallerArgumentCollection.OUTPUT_BLOCK_LOWER_BOUNDS + " refers to GVCF blocks," +
+                    " so reference confidence mode (" + AssemblyBasedCallerArgumentCollection.EMIT_REFERENCE_CONFIDENCE_LONG_NAME +
+                    ") must be specified.");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: You included this check twice here.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 3]
```diff
@@ -371,6 +372,43 @@ else if (!vc.getAlternateAllele(0).equals(Allele.NON_REF_ALLELE)){      //there
         }
     }
 
+    /*
+     * Test that in VCF mode we're consistent with past GATK4 results
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Comment does not match the test. Should be something like "Test the `--floor-blocks` argument."

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 4]
```diff
@@ -371,6 +372,43 @@ else if (!vc.getAlternateAllele(0).equals(Allele.NON_REF_ALLELE)){      //there
...
+    public void testFloorGVCFBlocks(final String inputFileName, final String referenceFileName) throws Exception {
+        Utils.resetRandomGenerator();
+
+        final File output = createTempFile("testFloorGVCFBlocks", ".vcf");
+
+        final List<String> requestedGqBands = Arrays.asList("10","20","30","40","50","60");
+
+        final ArgumentsBuilder args = new ArgumentsBuilder().addInput(new File(inputFileName))
+                .addReference(new File(referenceFileName))
+                .addInterval(new SimpleInterval("20:10000000-10100000"))
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: This interval might be unnecessarily large for this test -- could you shrink it down?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AnnotationUtils.java [hunk 5]
```diff
@@ -50,16 +50,7 @@ public static String encodeAnyASList( final List<?> somethingList) {
...
             return true;
         }
         return false;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AnnotationUtils.java]: `return annotation instanceof AlleleSpecificAnnotation;` would be more compact

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerArgumentCollection.java [hunk 6]
```diff
@@ -107,11 +108,12 @@ public ReadThreadingAssembler createReadThreadingAssembler() {
     public SmithWatermanAligner.Implementation smithWatermanImplementation = SmithWatermanAligner.Implementation.JAVA;
 
     /**
-     * (BETA feature) The reference confidence mode makes it possible to emit a per-bp or summarized confidence estimate for a site being strictly homozygous-reference.
-     * This is similar to the HaplotypeCaller reference confidence/GVCF mode. See https://software.broadinstitute.org/gatk/documentation/article.php?id=4017 for information about GVCFs.
+     * The reference confidence mode makes it possible to emit a per-bp or summarized confidence estimate for a site being strictly homozygous-reference.
+     * See https://software.broadinstitute.org/gatk/documentation/article.php?id=4017 for information about GVCFs.
+     * This is a BETA FEATURE for Mutect2 similar to the HaplotypeCaller reference confidence/GVCF mode.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerArgumentCollection.java]: Recommend rewording slightly to put "Mutect2" at the start of the sentence before the word "BETA": "For Mutect2, this is a BETA feature that functions similarly to the HaplotypeCaller reference confidence/GVCF mode."

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerArgumentCollection.java [hunk 7]
```diff
@@ -107,11 +108,12 @@ public ReadThreadingAssembler createReadThreadingAssembler() {
...
      */
     @Advanced
-    @Argument(fullName="emit-ref-confidence", shortName="ERC", doc="(BETA feature) Mode for emitting reference confidence scores", optional = true)
+    @Argument(fullName=EMIT_REFERENCE_CONFIDENCE_LONG_NAME, shortName="ERC", doc="Mode for emitting reference confidence scores (BETA for Mutect2)", optional = true)
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerArgumentCollection.java]: `(BETA for Mutect2)` -> `(For Mutect2, this is a BETA feature)`

**Comment by ldgauthier** [comment]
Sorry, that was sloppy.  I expect that now it builds.  Addressed all the review comments.

**Comment by droazen** [comment]
Looks like travis failed with an out-of-memory error (exit code 137) in gradle. @lbergelson was investigating this error earlier, and we thought we'd fixed it...

**Pull request merged**